### PR TITLE
Use `cmdutils.ConfigFileLoader`

### PR DESCRIPTION
**Pull request by errordeveloper**
### Description

Start using `cmdutils.ConfigFileLoader` where possible.

### Checklist

- [x] Code compiles correctly (i.e `make build`)
- [x] All unit tests passing (i.e. `make test`)
- [ ] All integration tests passing (i.e. `make integration-test`)

### CODE REVIEW

File: pkg/ctl/create/cluster.go [hunk 1]
```diff
@@ -103,69 +103,40 @@ func createClusterCmd(g *cmdutils.Grouping) *cobra.Command {
...
-		if meta.Region == "" {
-			return fmt.Errorf("metadata.region must be set")
-		}
-
-		p.Region = meta.Region
-
+	cfgLoader := cmdutils.NewClusterConfigLoader(p, cfg, clusterConfigFile, cmd)
+
+	cfgLoader.FlagsIncompatibleWithConfigFile.Insert(
+		"tags",
```
Comment (errordeveloper) [inline pkg/ctl/create/cluster.go]: `name`, `version` and `region` are already included (see #665).

File: pkg/ctl/create/cluster.go [hunk 2]
```diff
@@ -176,11 +147,11 @@ func doCreateCluster(p *api.ProviderConfig, cfg *api.ClusterConfig, nameArg stri
 
 		skipNodeGroupsIfRequested(cfg)
 
-		if err := CheckEachNodeGroup(ngFilter, cfg, NewNodeGroupChecker); err != nil {
-			return err
-		}
-	} else {
-		// validation and defaulting specific to when --config-file is unused
+		return CheckEachNodeGroup(ngFilter, cfg, NewNodeGroupChecker)
```
Comment (errordeveloper) [inline pkg/ctl/create/cluster.go]: It's a callback, so we can return this way

File: pkg/ctl/create/nodegroup.go [hunk 3]
```diff
@@ -49,7 +49,7 @@ func createNodeGroupCmd(g *cmdutils.Grouping) *cobra.Command {
 		fs.StringVar(&cfg.Metadata.Name, "cluster", "", "name of the EKS cluster to add the nodegroup to")
 		cmdutils.AddRegionFlag(fs, p)
 		cmdutils.AddVersionFlag(fs, cfg.Metadata, `for nodegroups "auto" and "latest" can be used to automatically inherit version from the control plane or force latest`)
-		fs.StringVarP(&clusterConfigFile, "config-file", "f", "", "load configuration from a file")
+		cmdutils.AddConfigFileFlag(&clusterConfigFile, fs)
```
Comment (errordeveloper) [inline pkg/ctl/create/nodegroup.go]: A thing we missed earlier...

File: pkg/ctl/create/cluster.go [hunk 4]
```diff
@@ -204,11 +175,10 @@ func doCreateCluster(p *api.ProviderConfig, cfg *api.ClusterConfig, nameArg stri
 
 			return nil
 		})
-		if err != nil {
-			return err
-		}
 	}
 
+	meta := cfg.Metadata
```
Comment (errordeveloper) [inline pkg/ctl/create/cluster.go]: TODO: actually call `cfgLoader.Load()`

File: pkg/ctl/delete/nodegroup.go [hunk 5]
```diff
@@ -56,31 +56,36 @@ func deleteNodeGroupCmd(g *cmdutils.Grouping) *cobra.Command {
...
+		}
 
-	if ng.Name != "" && nameArg != "" {
-		return cmdutils.ErrNameFlagAndArg(ng.Name, nameArg)
+		return nil
 	}
 
-	if nameArg != "" {
-		ng.Name = nameArg
+	if err := cfgLoader.Load(); err != nil {
```
Comment (errordeveloper) [inline pkg/ctl/delete/nodegroup.go]: TODO: I think this one actually needs much more work, will leave it out for another PR

File: pkg/ctl/create/cluster.go [hunk 6]
```diff
@@ -103,69 +103,40 @@ func createClusterCmd(g *cmdutils.Grouping) *cobra.Command {
...
+		"external-dns-access",
+		"full-ecr-access",
+		"storage-class",
+		"vpc-private-subnets",
+		"vpc-public-subnets",
+		"vpc-cidr",
+		"vpc-from-kops-cluster",
+	)
+
+	cfgLoader.ValidateWithConfigFile = func() error {
```
Comment (martina-if) [inline pkg/ctl/create/cluster.go]: I think I would move this validation functionality into the ClusterConfigLoader. It should be the config loader the one who knows how to validate stuff, IMO, not the caller. I should be able to delegate everything related to configuration loader to an "entity" that does it for me. Otherwise it looks like a leaky abstraction to me. What do you think?

Comment (errordeveloper) [reply pkg/ctl/create/cluster.go]: I agree, however currently the validation logic differers slightly in each of the commands, hence it was does as a custom callback. I agree that it's leaky at present.
I've tried to do something like `cfgLoader.ValidateWithConfigFile = cmdutils.ValidateCreateClusterWithConfigFile`, but wasn't sure what exactly was the benefit of that.
I suppose mapping commands to validator implementations inside of `cmdutils.ClusterConfigLoader` maybe an okay thing to do and we could move the lists of flags there also, as the bulk of the actually flags is already defined in `cmdutils` package.

Comment (errordeveloper) [reply pkg/ctl/create/cluster.go]: We could do something like
```
var validatorsWithConfigFile = map[string]func() error{
	"eksctl create cluster": func() error {
		// do stuff
		return nil
	},
}

// NewClusterConfigLoader constructs a standard loader
func NewClusterConfigLoader(provider *api.ProviderConfig, spec *api.ClusterConfig, clusterConfigFile string, cmd *cobra.Command) *ClusterConfigLoader {
	nilValidatorFunc := func() error { return nil }

	logger.Debug("cmd.CommandPath = %s", cmd.CommandPath())
	l := &ClusterConfigLoader{
		provider: provider,
		path:     clusterConfigFile,
		cmd:      cmd,

		Spec: spec,

		ValidateWithConfigFile:             nilValidatorFunc,
		FlagsIncompatibleWithConfigFile:    sets.NewString("name", "region", "version"),
		ValidateWithoutConfigFile:          nilValidatorFunc,
		FlagsIncompatibleWithoutConfigFile: sets.NewString(),
	}

	cmdPath := cmd.CommandPath()

	if validatorFunc, ok := validatorsWithConfigFile[cmdPath]; ok {
		l.ValidateWithConfigFile = validatorFunc
	}

	return l
}
```

Comment (martina-if) [reply pkg/ctl/create/cluster.go]: I think this would be the classic example of one interface with multiple implementations. So we could have one interface ConfigLoader, and then CreateClusterConfigLoader whose validation methods are specific for that. The rest of the common code, `Load()` and `LoadMetadata` would be delegated to the common config loader.

Comment (errordeveloper) [reply pkg/ctl/create/cluster.go]: Yeah, definitely a good idea! Will do that, thanks :)

**Comment by errordeveloper** [comment]
Since we are back to improving the loaded, I'm gonna open another PR
